python/src/aiconfig/editor/client/src/components/EditorContainer.tsx

```diff
@@ -665,9 +669,45 @@ export default function EditorContainer({
     return () => window.removeEventListener("keydown", saveHandler);
   }, [onSave]);
 
+  // Server heartbeat, check every 3s to show error if server is down
+  const getServerStatusCallback = callbacks.getServerStatus;
+  useEffect(() => {
+    if (!getServerStatusCallback) {
+      return;
+    }
+
+    const interval = setInterval(async () => {
+      try {
+        const res = await getServerStatusCallback();
+        setServerStatus(res.status);
+      } catch (err: unknown) {
+        setServerStatus("ERROR");
+      }
+    }, 3000);
+
+    return () => clearInterval(interval);
+  }, [getServerStatusCallback]);
+
   return (
     <AIConfigContext.Provider value={contextValue}>
       <Notifications />
+      {serverStatus !== "OK" && (
+        <>
+          {/* // Simple placeholder block div to make sure the banner does not overlap page contents until scrolling past its height */}
```

Comment:
This is actually incredible. Very few people know the small nuances like this and even less actually document it well so others know it easily!

Reply:
Lol. There's probably a cleaner way to do it with css (and/or js, to calculate alert height) but this is my "if it works..." solution